libs/tlv-derive/src/lib.rs

```diff
@@ -0,0 +1,152 @@
+    let entities = match data {
+        syn::Data::Struct(data) => match data.fields {
+            syn::Fields::Named(fields) => fields
+                .named
+                .into_iter()
+                .map(|f| parse_field_tlv_entity(&f).unwrap())
+                .collect::<Vec<_>>(),
+            _ => panic!("only named fields are supported"),
+        },
+        _ => panic!("only structs are supported"),
+    };
+    let tags = entities
+        .iter()
+        .map(|e| match e.tag {
+            tlv::Tag::U8(tag) => quote! { tlv::Tag::U8(#tag) },
+            tlv::Tag::U16(tag) => quote! { tlv::Tag::U16(#tag) },
+        })
+        .collect::<Vec<_>>();
+    let idents = entities.iter().map(|e| &e.ident).collect::<Vec<_>>();
+    let gen = quote! {
+        impl tlv::Encode for #name {
+            fn encode<W>(&self, writer: &mut W) -> std::io::Result<()>
+            where
+                W: std::io::Write,
+            {
+                #(tlv::value::encode_tagged(#tags, &self.#idents, writer)?;)*
+                Ok(())
+            }
+
+            fn encoded_length(&self) -> std::io::Result<tlv::Length> {
+                let mut length = tlv::Length::new(0);
+                #(length += tlv::value::length_tagged(#tags, &self.#idents)?;)*
+                Ok(length)
+            }
+        }
+    };
+    gen.into()
+}
+
+/// Generates an implementation of the `Decode` trait for the annotated struct.
+/// The implementation will decode the struct from a TLV format.
+#[proc_macro_derive(Decode, attributes(tlv))]
+pub fn decode_derive(input: TokenStream) -> TokenStream {
+    // Construct a representation of Rust code as a syntax tree
+    // that we can manipulate
+    let DeriveInput { ident, data, .. } = parse_macro_input!(input);
+
+    let name = ident;
+    let entities = match data {
+        syn::Data::Struct(data) => match data.fields {
+            syn::Fields::Named(fields) => fields
+                .named
+                .into_iter()
+                .map(|f| parse_field_tlv_entity(&f).unwrap())
+                .collect::<Vec<_>>(),
+            _ => panic!("only named fields are supported"),
+        },
+        _ => panic!("only structs are supported"),
+    };
+    let tags = entities.iter().map(|e| match e.tag {
+        tlv::Tag::U8(tag) => quote! { tlv::Tag::U8(#tag) },
+        tlv::Tag::U16(tag) => quote! { tlv::Tag::U16(#tag) },
+    });
+    let idents = entities.iter().map(|e| &e.ident).collect::<Vec<_>>();
+    let gen = quote! {
+        impl tlv::Decode for #name {
+            fn decode<R>(reader: &mut R) -> std::io::Result<(Self, usize)>
+            where
+                R: std::io::Read
+            {
+                let __tlv_decode_read: usize = 0;
+                #(
+                    let (#idents, __tlv_decode_read) = match tlv::value::decode_tagged(#tags, reader)? {
+                        (value, read) => (value, read + __tlv_decode_read),
+                    };
```

Comment:
This takes advantage of the fact that Rust lets you redeclare a variable with the same name, shadowing the previous declaration, using the old one in the initializer.